Performance.md

```diff
@@ -0,0 +1,290 @@
+Each project has its own `tsconfig.json` that has dependencies on other projects.
+This can be helpful to avoid loading too many files in a single compilation, and also makes certain codebase layout strategies easier to put together.
+
+There are some very basic ways of breaking up a codebase into projects.
+As an example, one might be a program with a project for the client, a project for the server, and a project that's shared between the two.
+
+```
+              ------------
+              |          |
+              |  Shared  |
+              ^----------^
+             /            \
+            /              \
+------------                ------------
+|          |                |          |
+|  Client  |                |  Server  |
+-----^------                ------^-----
+```
+
+Tests can also be broken into their own project.
+
+```
+              ------------
+              |          |
+              |  Shared  |
+              ^-----^----^
+             /      |     \
+            /       |      \
+------------  ------------  ------------
+|          |  |  Shared  |  |          |
+|  Client  |  |  Tests   |  |  Server  |
+-----^------  ------------  ------^-----
+     |                            |
+     |                            |
+------------                ------------
+|  Client  |                |  Server  |
+|  Tests   |                |  Tests   |
+------------                ------------
+```
+
+You can read up more about project references here. <!-- TODO -->
+
+# Configuring `tsconfig.json`
+
+Within a `tsconfig.json`, there are two ways to specify files in a project:
+
+* the `files` list
+* the `include` and `exclude` lists
+
+The primary difference between the two is that `files` expects a list of file paths to source files, and `include/`exclude` use globbing patterns to match against files.
+
+While specifying `files` will allow TypeScript to quickly load up files up directly, it can be cumbersome if you have many files in your project without just a few top-level entry-points.
+Additionally, it's easy to forget to add new files to your `tsconfig.json`, which means that you might end up with strange editor behavior where those new files are incorrectly analyzed.
+All this can be cumbersome.
+
+`include`/`exclude` help avoid needing to specify these files, but at a cost: files must be discovered by walking through included directories.
+When running through a *lot* of folders, this can slow compilations down.
+Additionally, sometimes a compilation will include lots of unnecessary `.d.ts` files and test files, which can increase compilation time and memory overhead.
+Finally, while `exclude` has some reasonable defaults, certain configurations like mono-repos mean that a "heavy" folders like `node_modules` can still end up being included.
+
+For best practices, we recommend the following:
+
+* Specify only source folders in your project.
```

Comment:
I'm not sure what this means.  Above we were talking about files.  Is this a way to avoid globbing?